### ISSUE TITLE
Feature request: C# extensibility points at runtime

### ISSUE THREAD

**Issue by koolraap** [Azure/autorest#1021]
I'd quite like to be able to intercept all method calls made to an Autorest generated method. I can do this with an Autofac (Castle Windsor) interceptor or with a wrapper class but it'd be nice if there was a Pre/Post handler.

My motivation for this is to use the extensibility points to make the generated proxies self-authenticating (and re-authenticating) and to set various standard headers per call.

Prerequisite: implement #452.

I haven't though through the details of this in any depth, but the requirements are approximately:
1. Ability to register a PreApiCall handler per class. Bonus points for offering delegate and Func<> overloads.
2. Ability to register a PostApiCall handler per class <-- this is really an attempt to implement #931 by stealth.
3. Change code generation for *WithHttpMessagesAsync methods so the first thing it does is call the PreCall handler with an IServiceClientWithCredentials (whatever, see #452) and access to the customHeaders and cancellationToken parameters and anything else that we can get our hands on, Verb, Url etc.
4. As 3, but immediately after we have the result call the PostApiCall handler with all the details (including the body!).

Or, if there's a really easy way to meet my objectives that I've missed, I'd love to hear it.

**Comment by fearthecowboy** [comment]
Howdy!

In our planning for driving towards a stable '1.0' release, I'm marking this issue as 'deferred' :zzz: and we're going to review it during the post-1.0 planning cycle. 

_It's not to say that we're not going to work on it, or that this isn't not important, but at the moment, we're picking and choosing the stuff we **must** do before 1.0._ :horse_racing:  :horse_racing:  :horse_racing: 

We'll make sure we pick this back up at that point. :tada:
